feat: resolve resource links [DANTE-852]

Silhoue opened the pull request:
Add support for client-side ResourceLinks resolution (basic implementation).

`.getEntry()` and `.getEntries()` functions now resolve references to entries from other spaces as they would for regular references, respecting existing `resolveLinks`, `removeUnresolved` and `include` parameters.

Known issues/follow-up tasks are documented in the ticket.

BasKiers commented on lib/utils/resource-links.js:
```diff
@@ -0,0 +1,60 @@
+function isResourceLink (field) {
+  return field && field.sys && field.sys.type === 'ResourceLink'
+}
```
I would make this check a bit more resilient to prevent bugs when the customer is relying on this SDK version and we release new `ResourceLink` types.
```suggestion
function isContentfulEntryResourceLink (field) {
  return field && field.sys && field.sys.type === 'ResourceLink' && field.sys.linkType === 'Contentful:Entry'
}
```

BasKiers commented on lib/utils/resource-links.js:
```diff
@@ -0,0 +1,60 @@
+
+function parseResourceLink (resourceLink) {
+  const urn = resourceLink.sys.urn
+  // For now just assume every resourceLink has type Contentful:Entry
+  const [, spaceId, , entryId] = urn.split('/')
```
This was written pretty hastily (by me), is it worth the effort to use a better urn parse function?

Silhoue replied:
Since we've already checked that this is a `Contentful:Entry` ResourceLink it should be fine. If we were to change something in the crn pattern it will stop working anyway in the [resolve-response](https://github.com/contentful/contentful-resolve-response/blob/b44ce875c33048d2c19c9a7ca7dd341bd1a237da/index.js#L63).

BasKiers commented on lib/utils/resource-links.js:
```diff
@@ -0,0 +1,60 @@
+
+function getResourceLinkFields (data) {
+  const items = [].concat(...(data.items || []), ...Object.values(data.includes || {}).flat())
+  const fields = items.flatMap(({ fields }) => Object.values(fields).flat())
+  return fields.filter(isResourceLink)
```
This code breaks if the `locale=*` parameter is set as it assumes the field response does not include the localisation map

Silhoue replied:
Good point, I'll look into this

BasKiers commented on lib/utils/resource-links.js:
```diff
@@ -0,0 +1,60 @@
+
+  const entryIds = entries.map(({ entryId }) => entryId)
+
+  let response = { items: [] }
```
Does not need to be a `let` but can be instantiated as a `const` on line 36
```suggestion
```

BasKiers commented on lib/utils/resource-links.js:
```diff
@@ -0,0 +1,60 @@
+  let response = { items: [] }
+  const uniqueEntryIds = [...new Set(entryIds)]
+  response = await apiInstance.getEntries({ 'sys.id[in]': uniqueEntryIds.join(','), include }).catch(() => ({}))
+  return response.items || []
```
We should probably not only return the `items` but also the `includes` from the response (for assets and same-space Links). This would probably mean changing the return type which will require some refactoring in `addResourceLinksToIncludes`
```suggestion
  return {...(response.includes || {}), Entry: {...((response.includes && response.includes.Entry) || []), ...(response.items || [])}
```

BasKiers replied:
I was incorrect, I forgot that the `apiInstance` will call it's own `wrapEntryCollection` function to copy the includes into the field key

BasKiers commented on lib/utils/resource-links.js:
```diff
@@ -0,0 +1,60 @@
+  return response.items || []
+}
+
+export async function addResourceLinksToIncludes (data, extraApis = {}, include = 1) {
```
FYI: This function is not written recursively. You'll only be able to resolve 1 layer of x-space ResourceLinks.

Silhoue replied:
It is, `addResourceLinksToIncludes` calls `getEntitiesFromSpace` which calls `getEntries` which calls `addResourceLinksToIncludes` again

BasKiers replied:
You are right. I made the same mistake in my other comment where I forgot that every api instance is doing its own recursive resolving

Silhoue replied:
I added a test case for it

luizfonseca commented on test/integration/resource-links-test.js:
```diff
@@ -0,0 +1,137 @@
+ */
+
+const client = contentful.createClient({
+  accessToken: '2O_YlI7Spwm16XXw-VZw9vxIIlUOeHLuzWwKXn3KMvo',
+  space: 'z788fz497ijs',
+  additionalTokens: {
+    mbfd18am5xhd: 'tZO0vih_LPefnfNhCDARaRjfiSNXLeJMYRvFe7U-6Yk',
+    p9sqh626r1fo: 'CagovFaZwA7Dta0chjKBU2iU4ASND0TLgCgQyep-Xnk'
+  }
```
are those real tokens or mocked ones?

Silhoue replied:
Real CDA tokens to the testing space. This is how integration tests are currently working here.

luizfonseca commented:
Could you add a brief description to the PR indicating what this is adding and it works?

Maybe some points to clarify:

- Does it keep resolving cross-space links until it finds nothing?
- Can the user add a limit?

BasKiers commented on lib/utils/resource-links.js:
```diff
@@ -0,0 +1,64 @@
+
+  const entryIds = entries.map(({ entryId }) => entryId)
+  const uniqueEntryIds = [...new Set(entryIds)]
+  const response = await apiInstance.getEntries({ 'sys.id[in]': uniqueEntryIds.join(','), include, locale }, { wrapEntryCollection: false }).catch(() => ({}))
+  return getEntries(response)
```
I believe this will cause `Asset` fields from the `includes` of the `ResourceLink` object to be discarded

BasKiers replied:
I would suggest putting `items` in the `includes.Entry` array and returning the `includes` object as a whole. In `addResourceLinksToIncludes` you can rewrite the merging logic to merge all `includes` keys e.g.
```javascript
const resourceLinkEntries = (await Promise.all(resourceLinkResponses)).flat() // {[type: String]: Record<string, any>[]}[]
const includesObjects = [...resourceLinkEntries, data.includes || {}]
const includesKeys = new Set(includesObjects.flatMap(({includes = {}}) => Object.keys(includes))
return {
  ...data,
  includes: Array.from(...includesKeys).reduce((includes, key) => ({
    ...includes,
    [key]: includesObjects.flatMap((incl) => incl[key] || [])
  }), {})
}
```

luizfonseca commented:
You have to update these types as well: https://github.com/contentful/contentful.js/blob/6451a5c286f2243d233008cfb49b5eae1682dd9d/index.d.ts#L16

luizfonseca commented on lib/utils/resource-links.js:
```diff
@@ -0,0 +1,64 @@
+    ...data,
+    includes: {
+      ...data.includes,
+      Entry: [...((data.includes && data.includes.Entry) || []), ...resourceLinkEntries]
```
Q: similar to @BasKiers  comment, I am querying using the client lib but I don't get any of the includes of the other request (e.g. assets etc)

Is it something we need to support now or a follow-up?

Silhoue replied:
I pushed a fix for including the assets. Does that fix the issue you noticed or is it something else?

luizfonseca replied:
Just checked; fixed 👍🏽

Cyberxon commented on ADVANCED.md:
```diff
@@ -62,6 +62,29 @@ The link resolution is applied to one level deep by default. If you need it to b
 
+### Links to other spaces (beta)
+
+`getEntry` and `getEntries` endponts can now resolve links to entries from other spaces as they do for regular links. The `resolveLinks`, `removeUnresolved` and `include` parameters affect both types of links.
```
endpoints*

Cyberxon commented on ADVANCED.md:
```diff
@@ -62,6 +62,29 @@ The link resolution is applied to one level deep by default. If you need it to b
+  accessToken: "<your-access-token>",
+  space: "<your-space-id>",
+  additionalTokens: {
+    "<other-space-id>": "<other-access-token>",
```
maybe make it clear that the access token is that space access token by rephrasing to `<other-space-access-token>`

Cyberxon commented on ADVANCED.md:
````diff
@@ -62,6 +62,29 @@ The link resolution is applied to one level deep by default. If you need it to b
+```js
+const contentful = require("contentful");
+const client = contentful.createClient({
+  accessToken: "<your-access-token>",
````
and here to `<current-space-access-token>`

Silhoue commented:
> Could you add a brief description to the PR indicating what this is adding and how it works?
Updated the PR description and added a section in the docs

luizfonseca reviewed (approved):
Besides the update on `index.d.ts`, no big comments from my side. 🚀

Cyberxon commented on README.md:
```diff
@@ -249,6 +249,13 @@ contentful.createClient({
+      <td><code>additionalTokens</code></td>
+      <td><code>{}</code></td>
+      <td>
+        Additional access tokens, used for resolving links to other spaces. The expected format is <code>{ spaceId: accessToken, ... }</code>
```
Also here in the readme maybe make it clear that each spaceId should have their respective access token

Cyberxon commented on lib/contentful.js:
```diff
@@ -75,20 +77,36 @@ export function createClient (params) {
+    })
+  }
+
+  const contentfulApiBySpaceId = Object.entries(config.additionalTokens).reduce((result, [spaceId, accessToken]) => {
```
One improvement I can think of is comparing this access token to the access token provided for the main space and not calling the function if it is the case

Cyberxon reviewed (commented):
Great work! 💯  Just some minor comments

Silhoue commented on lib/contentful.js:
```diff
@@ -75,20 +77,36 @@ export function createClient (params) {
-    http,
-    getGlobalOptions
-  })
+  Object.values(contentfulApiBySpaceId).forEach(contentfulApi => { contentfulApi.extraApis = contentfulApiBySpaceId })
```
I renamed it to `additionalApis`, similar to `additionalTokens`.
